File: modules/ml-analyzer.js
/**
 * Use machine learning techniques to detect code smells
 */
const fs = require('fs');
const path = require('path');

class MLAnalyzer {
  constructor(metrics) {
    this.metrics = metrics;
    this.features = {};
  }
  
  /**
   * Extract code features for ML-based analysis
   */
  extractFeatures(filePath, content, ast) {
    // Basic metrics
    const lines = content.split('\n').length;
    const chars = content.length;
    const commentLines = (content.match(/\/\/.*/g) || []).length + 
                        ((content.match(/\/\*[\s\S]*?\*\//g) || [])
                          .join('\n').split('\n').length);
    
    // AST-based metrics
    let maxNestingLevel = 0;
    let currentNestingLevel = 0;
    let functionCount = 0;
    let longFunctionCount = 0;
    let complexConditionCount = 0;
    
    traverse(ast, {
      enter(node) {
        // Track nesting level
        if (node.type === 'BlockStatement' || 
            node.type === 'ObjectExpression' || 
            node.type === 'ArrayExpression') {
          currentNestingLevel++;
          maxNestingLevel = Math.max(maxNestingLevel, currentNestingLevel);
        }
        
        // Count functions
        if (node.type === 'FunctionDeclaration' || 
            node.type === 'FunctionExpression' || 
            node.type === 'ArrowFunctionExpression') {
          functionCount++;
          
          // Check for long functions
          if (node.body && node.body.type === 'BlockStatement' && 
              node.body.body && node.body.body.length > 15) {
            longFunctionCount++;
          }
        }
        
        // Check for complex conditions
        if ((node.type === 'LogicalExpression' || node.type === 'BinaryExpression') &&
            node.left && node.right &&
            (node.left.type === 'LogicalExpression' || node.left.type === 'BinaryExpression' ||
             node.right.type === 'LogicalExpression' || node.right.type === 'BinaryExpression')) {
          complexConditionCount++;
        }
      },
      exit(node) {
        if (node.type === 'BlockStatement' || 
            node.type === 'ObjectExpression' || 
            node.type === 'ArrayExpression') {
          currentNestingLevel--;
        }
      }
    });
    
    // Store features
    this.features[filePath] = {
      lines,
      chars,
      commentLines,
      commentRatio: commentLines / lines,
      maxNestingLevel,
      functionCount,
      longFunctionCount,
      longFunctionRatio: functionCount > 0 ? longFunctionCount / functionCount : 0,
      complexConditionCount,
      avgCharsPerLine: lines > 0 ? chars / lines : 0,
      timestamp: Date.now()
    };
    
    return this.features[filePath];
  }
  
  /**
   * Detect abnormal code using statistical methods
   */
  detectAbnormalCode() {
    const files = Object.keys(this.features);
    if (files.length < 5) return []; // Need enough samples
    
    // Calculate mean and standard deviation for each metric
    const metrics = ['maxNestingLevel', 'longFunctionRatio', 'complexConditionCount', 'commentRatio'];
    const stats = {};
    
    metrics.forEach(metric => {
      const values = files.map(file => this.features[file][metric]);
      const mean = values.reduce((sum, val) => sum + val, 0) / values.length;
      const variance = values.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / values.length;
      const stdDev = Math.sqrt(variance);
      
      stats[metric] = { mean, stdDev };
    });
    
    // Detect outliers (z-score > 2)
    const outliers = [];
    
    files.forEach(file => {
      const feature = this.features[file];
      const issues = [];
      
      metrics.forEach(metric => {
        const zScore = Math.abs((feature[metric] - stats[metric].mean) / stats[metric].stdDev);
        
        if (zScore > 2) {
          issues.push({
            metric,
            value: feature[metric],
            mean: stats[metric].mean,
            zScore
          });
        }
      });
      
      if (issues.length > 0) {
        outliers.push({
          file,
          issues,
          score: issues.reduce((sum, issue) => sum + issue.zScore, 0) / issues.length
        });
      }
    });
    
    // Sort by outlier score
    outliers.sort((a, b) => b.score - a.score);
    
    // Update metrics with ML findings
    this.metrics.codeQuality.mlAnalysis = {
      outliers,
      stats
    };
    
    return outliers;
  }

  // Add this method to the MLAnalyzer class

  /**
   * Generate embeddings for RAG documents
   */
  async generateEmbeddingsForRagDocuments(ragDir) {
    console.log("Generating embeddings for RAG documents...");
    
    // Check if embedding capability is available
    if (!this.tensorflowLoaded) {
      console.warn("TensorFlow not available for embedding generation");
      return;
    }
    
    const embeddingsDir = path.join(ragDir, 'embeddings');
    fs.ensureDirSync(embeddingsDir);
    
    // Get all markdown files recursively
    const files = this.getAllMarkdownFiles(ragDir);
    console.log(`Found ${files.length} documents for embedding generation`);
    
    // Generate embeddings for each document
    const embeddings = {};
    
    for (const file of files) {
      try {
        // Skip embedding files
        if (file.includes('embeddings/')) continue;
        
        // Read file content
        const content = fs.readFileSync(file, 'utf8');
        
        // Generate embedding
        const embedding = await this.generateDocumentEmbedding(content);
        
        // Store with relative path as key
        const relativePath = path.relative(ragDir, file);
        embeddings[relativePath] = embedding;
        
        console.log(`Generated embedding for ${relativePath}`);
      } catch (err) {
        console.error(`Error generating embedding for ${file}:`, err.message);
      }
    }
    
    // Save all embeddings
    fs.writeFileSync(
      path.join(embeddingsDir, 'document_embeddings.json'),
      JSON.stringify(embeddings, null, 2)
    );
    
    // Generate metadata file
    const metadata = {
      created: new Date().toISOString(),
      documentCount: Object.keys(embeddings).length,
      embeddingDimensions: Object.values(embeddings)[0]?.length || 0,
      model: 'universal-sentence-encoder'
    };
    
    fs.writeFileSync(
      path.join(embeddingsDir, 'metadata.json'),
      JSON.stringify(metadata, null, 2)
    );
    
    console.log(`Generated embeddings for ${Object.keys(embeddings).length} documents`);
    return embeddings;
  }

  /**
   * Generate embedding for a document
   */
  async generateDocumentEmbedding(text) {
    // This is a simplified version - you'd use actual embedding model here
    if (!this.use) {
      try {
        // Load Universal Sentence Encoder
        this.use = await require('@tensorflow-models/universal-sentence-encoder').load();
      } catch (err) {
        console.error("Error loading Universal Sentence Encoder:", err.message);
        return null;
      }
    }
    
    // Generate embedding
    try {
      const embeddings = await this.use.embed(text);
      const embeddingArray = await embeddings.array();
      return embeddingArray[0]; // First item in batch
    } catch (err) {
      console.error("Error generating document embedding:", err.message);
      return null;
    }
  }

  /**
   * Get all markdown files recursively
   */
  getAllMarkdownFiles(dir) {
    let results = [];
    const list = fs.readdirSync(dir);
    
    list.forEach(file => {
      const fullPath = path.join(dir, file);
      const stat = fs.statSync(fullPath);
      
      if (stat.isDirectory()) {
        results = results.concat(this.getAllMarkdownFiles(fullPath));
      } else if (file.endsWith('.md')) {
        results.push(fullPath);
      }
    });
    
    return results;
  }
}

module.exports = MLAnalyzer;

File: src/modules/ml_analyzer.rs
use std::path::{Path, PathBuf};
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};
use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use log::info;

/// Represents an Abstract Syntax Tree (AST) node
#[derive(Debug, Clone)]
pub enum AstNode {
    // This is a simplified representation - would need to be expanded
    // based on the actual AST structure used in your project
    Block(Vec<Box<AstNode>>),
    Function(FunctionNode),
    Condition(ConditionNode),
    Object(Vec<(String, Box<AstNode>)>),
    Array(Vec<Box<AstNode>>),
    Other(String),
}

#[derive(Debug, Clone)]
pub struct FunctionNode {
    pub name: Option<String>,
    pub params: Vec<String>,
    pub body: Box<AstNode>,
}

#[derive(Debug, Clone)]
pub struct ConditionNode {
    pub operator: String,
    pub left: Box<AstNode>,
    pub right: Box<AstNode>,
}

/// Code features extracted for ML analysis
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeFeatures {
    pub lines: usize,
    pub chars: usize,
    pub comment_lines: usize,
    pub comment_ratio: f64,
    pub max_nesting_level: usize,
    pub function_count: usize,
    pub long_function_count: usize,
    pub long_function_ratio: f64,
    pub complex_condition_count: usize,
    pub avg_chars_per_line: f64,
    pub timestamp: u64,
}

/// Outlier issue in code
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeIssue {
    pub metric: String,
    pub value: f64,
    pub mean: f64,
    pub z_score: f64,
}

/// Code outlier detected through ML analysis
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeOutlier {
    pub file: PathBuf,
    pub issues: Vec<CodeIssue>,
    pub score: f64,
}

/// Metric statistics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricStats {
    pub mean: f64,
    pub std_dev: f64,
}

/// ML analysis results
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MlAnalysisResults {
    pub outliers: Vec<CodeOutlier>,
    pub stats: HashMap<String, MetricStats>,
}

/// Use machine learning techniques to detect code smells
pub struct MLAnalyzer<M> {
    metrics: M,
    features: HashMap<PathBuf, CodeFeatures>,
}

impl<M> MLAnalyzer<M> {
    /// Create a new ML analyzer with the given metrics
    pub fn new(metrics: M) -> Self {
        Self {
            metrics,
            features: HashMap::new(),
        }
    }
    
    /// Extract code features for ML-based analysis
    pub fn extract_features(&mut self, file_path: &Path, content: &str, ast: &AstNode) -> Result<&CodeFeatures> {
        // Basic metrics
        let lines = content.lines().count();
        let chars = content.len();
        
        // Count comment lines using a simple regex-based approach
        // This is a simplification - a real implementation would use the AST
        let single_line_comments = content.matches("//").count();
        let multi_line_comments = content.matches("/*").count();
        let comment_lines = single_line_comments + multi_line_comments;
        
        // AST-based metrics
        let mut max_nesting_level = 0;
        let mut function_count = 0;
        let mut long_function_count = 0;
        let mut complex_condition_count = 0;
        
        // Process AST to extract metrics
        self.traverse_ast(
            ast, 
            0, 
            &mut max_nesting_level,
            &mut function_count,
            &mut long_function_count,
            &mut complex_condition_count,
        );
        
        // Store features
        let features = CodeFeatures {
            lines,
            chars,
            comment_lines,
            comment_ratio: if lines > 0 { comment_lines as f64 / lines as f64 } else { 0.0 },
            max_nesting_level,
            function_count,
            long_function_count,
            long_function_ratio: if function_count > 0 { 
                long_function_count as f64 / function_count as f64 
            } else { 
                0.0 
            },
            complex_condition_count,
            avg_chars_per_line: if lines > 0 { chars as f64 / lines as f64 } else { 0.0 },
            timestamp: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_millis() as u64,
        };
        
        self.features.insert(file_path.to_path_buf(), features.clone());
        
        Ok(self.features.get(file_path).unwrap())
    }
    
    /// Helper method to traverse the AST and collect metrics
    fn traverse_ast(
        &self,
        node: &AstNode,
        current_nesting_level: usize,
        max_nesting_level: &mut usize,
        function_count: &mut usize,
        long_function_count: &mut usize,
        complex_condition_count: &mut usize,
    ) {
        match node {
            AstNode::Block(children) => {
                // Increase nesting level for blocks
                let new_nesting_level = current_nesting_level + 1;
                *max_nesting_level = (*max_nesting_level).max(new_nesting_level);
                
                // Process children
                for child in children {
                    self.traverse_ast(
                        child, 
                        new_nesting_level,
                        max_nesting_level,
                        function_count,
                        long_function_count,
                        complex_condition_count,
                    );
                }
            },
            AstNode::Function(function) => {
                // Count functions
                *function_count += 1;
                
                // Check for long functions
                if let AstNode::Block(ref children) = *function.body {
                    if children.len() > 15 {
                        *long_function_count += 1;
                    }
                }
                
                // Process function body
                self.traverse_ast(
                    &function.body, 
                    current_nesting_level,
                    max_nesting_level,
                    function_count,
                    long_function_count,
                    complex_condition_count,
                );
            },
            AstNode::Condition(condition) => {
                // Check for complex conditions
                let is_left_complex = matches!(*condition.left, AstNode::Condition(_));
                let is_right_complex = matches!(*condition.right, AstNode::Condition(_));
                
                if is_left_complex || is_right_complex {
                    *complex_condition_count += 1;
                }
                
                // Process condition parts
                self.traverse_ast(
                    &condition.left, 
                    current_nesting_level,
                    max_nesting_level,
                    function_count,
                    long_function_count,
                    complex_condition_count,
                );
                
                self.traverse_ast(
                    &condition.right, 
                    current_nesting_level,
                    max_nesting_level,
                    function_count,
                    long_function_count,
                    complex_condition_count,
                );
            },
            AstNode::Object(fields) => {
                // Increase nesting level for objects
                let new_nesting_level = current_nesting_level + 1;
                *max_nesting_level = (*max_nesting_level).max(new_nesting_level);
                
                // Process fields
                for (_, value) in fields {
                    self.traverse_ast(
                        value, 
                        new_nesting_level,
                        max_nesting_level,
                        function_count,
                        long_function_count,
                        complex_condition_count,
                    );
                }
            },
            AstNode::Array(items) => {
                // Increase nesting level for arrays
                let new_nesting_level = current_nesting_level + 1;
                *max_nesting_level = (*max_nesting_level).max(new_nesting_level);
                
                // Process items
                for item in items {
                    self.traverse_ast(
                        item, 
                        new_nesting_level,
                        max_nesting_level,
                        function_count,
                        long_function_count,
                        complex_condition_count,
                    );
                }
            },
            _ => {
                // Other nodes don't affect our metrics
            },
        }
    }
    
    /// Detect abnormal code using statistical methods
    pub fn detect_abnormal_code(&mut self) -> Vec<CodeOutlier> {
        let files: Vec<PathBuf> = self.features.keys().cloned().collect();
        if files.len() < 5 {
            info!("Not enough samples for ML analysis, need at least 5 files");
            return vec![];
        }
        
        // Calculate mean and standard deviation for each metric
        let metrics = vec![
            "max_nesting_level".to_string(),
            "long_function_ratio".to_string(),
            "complex_condition_count".to_string(),
            "comment_ratio".to_string(),
        ];
        
        let mut stats: HashMap<String, MetricStats> = HashMap::new();
        
        for metric in &metrics {
            let values: Vec<f64> = files.iter()
                .map(|file| {
                    let feature = &self.features[file];
                    match metric.as_str() {
                        "max_nesting_level" => feature.max_nesting_level as f64,
                        "long_function_ratio" => feature.long_function_ratio,
                        "complex_condition_count" => feature.complex_condition_count as f64,
                        "comment_ratio" => feature.comment_ratio,
                        _ => 0.0,
                    }
                })
                .collect();
            
            let mean = values.iter().sum::<f64>() / values.len() as f64;
            
            let variance = values.iter()
                .map(|val| (val - mean).powi(2))
                .sum::<f64>() / values.len() as f64;
            
            let std_dev = variance.sqrt();
            
            stats.insert(metric.clone(), MetricStats { mean, std_dev });
        }
        
        // Detect outliers (z-score > 2)
        let mut outliers: Vec<CodeOutlier> = Vec::new();
        
        for file in &files {
            let feature = &self.features[file];
            let mut issues: Vec<CodeIssue> = Vec::new();
            
            for metric in &metrics {
                let value = match metric.as_str() {
                    "max_nesting_level" => feature.max_nesting_level as f64,
                    "long_function_ratio" => feature.long_function_ratio,
                    "complex_condition_count" => feature.complex_condition_count as f64,
                    "comment_ratio" => feature.comment_ratio,
                    _ => 0.0,
                };
                
                let stat = &stats[metric];
                
                // Avoid division by zero
                if stat.std_dev > 0.0 {
                    let z_score = (value - stat.mean).abs() / stat.std_dev;
                    
                    if z_score > 2.0 {
                        issues.push(CodeIssue {
                            metric: metric.clone(),
                            value,
                            mean: stat.mean,
                            z_score,
                        });
                    }
                }
            }
            
            if !issues.is_empty() {
                let score = issues.iter().map(|issue| issue.z_score).sum::<f64>() / issues.len() as f64;
                
                outliers.push(CodeOutlier {
                    file: file.clone(),
                    issues,
                    score,
                });
            }
        }
        
        // Sort by outlier score
        outliers.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
        
        // Update metrics with ML findings
        // This would require accessing the appropriate field in the metrics struct
        // depending on how it's structured
        
        outliers
    }
    
    /// Helper method for embedding generation
    /// This would be implemented to work with any ML models you integrate
    pub async fn generate_document_embedding(&self, text: &str) -> Result<Vec<f32>> {
        // Simplified implementation - in a real scenario this would call
        // an actual embedding model or use a library like `rust-bert`
        
        // Generate a simple hash-based embedding as a placeholder
        let hash = text.chars().fold(0u32, |acc, c| {
            acc.wrapping_add(c as u32)
        });
        
        // Create a 512-dimensional embedding vector using the hash as a seed
        let embedding: Vec<f32> = (0..512)
            .map(|i| ((hash.wrapping_add(i as u32)) as f32 / u32::MAX as f32) * 2.0 - 1.0)
            .collect();
            
        Ok(embedding)
    }
}
